shared/src/balancer_event_handler.rs

```diff
@@ -0,0 +1,154 @@
+use crate::{
+    current_block::BlockRetrieving,
+    event_handling::{BlockNumber, EventHandler, EventIndex, EventStoring},
+    impl_event_retrieving,
+    maintenance::Maintaining,
+};
+use anyhow::{anyhow, Context, Result};
+use contracts::{
+    balancer_v2_vault::{
+        self, event_data::PoolRegistered as ContractPoolRegistered, Event as ContractEvent,
+    },
+    BalancerV2Vault,
+};
+use ethcontract::common::DeploymentInformation;
+use ethcontract::{dyns::DynWeb3, Event as EthContractEvent, EventMetadata, H160};
+use std::ops::RangeInclusive;
+use tokio::sync::Mutex;
+
+#[derive(Debug)]
+pub enum BalancerEvent {
+    PoolRegistered(PoolRegistered),
+}
+
+#[derive(Debug, Default)]
+pub struct PoolRegistered {
+    pub pool_id: [u8; 32],
+    pub pool_address: H160,
+    pub specialization: u8,
```

Comment:
Can you add comments to this struct? Specifically specialization is not intuitive to me (might suffice to link to the balancer docs explaining what this is)

Reply:
This is a good point. I would suggest:
```
pub struct PoolRegistered {
    pub pool_id: H256,
    pub pool_address: H160,
    pub specialization: PoolSpecialization,
}

#[repr(u8)]
pub enum PoolSpecialization {
    General = 0,
    MinimalSwapInfo = 1,
    TwoToken = 2,
}
```

Reply:
In theory, we could even copy [the documentation](https://github.com/balancer-labs/balancer-core-v2/blob/3b8c2fb5c75e987d63c4d44f9dd62351f43e153b/pkg/vault/contracts/interfaces/IVault.sol#L197-L212) from the Balancer repo to keep it close.

Reply:
Ok, so I have included the suggestions here. Note that I had to quite a bit of manual labor for this `PoolSpecialization`struct (i.e. manual implementation of `Debug` and an explicit `new` constructor which builds the enum from u8, but has to return a result because of Error when the value is too high).

Not sure if I am missing out on some sweet tricks here, but I guess thats all part of the safety guarantees!

I also, don't see the real benefit of using `H256` here for `pool_id` instead of `[u8; 32]` if we aren't going to make use of any functionality inherited from the fact that its an `H256`.